src/dynamics/omega_limit.lean

```diff
@@ -0,0 +1,454 @@
+      (λ he, h₂ t ht.2 (nonempty.mono
+        (inter_subset_inter_right _ (preimage_mono (inter_subset_right _ _))) he)) },
+  { intros y hy,
+    exact or.elim hy
+    (λ hy₂, omega_limit_mono_right _ _ (subset_union_left  _ _) hy₂)
+    (λ hy₂, omega_limit_mono_right _ _ (subset_union_right _ _) hy₂) },
+end
+
+-- ω-limits in compact spaces:
+
+/-- A set is eventually carried into any neighbourhood of its ω-limit:
+    for any neighbourhood v of ω f ϕ s, there exists u ∈ f such that
+    cl (ϕ u s) ⊆ v -/
+lemma eventually_subset_nhd_omega_limit [compact_space β]
+  {v : set β} (ho : is_open v) (hv : ω f ϕ s ⊆ v) :
+  ∃ u ∈ f, closure (image2 ϕ u s) ⊆ v :=
+begin
+  let j := λ u, (closure (image2 ϕ u s))ᶜ,
+  have hj₁ : ∀ u ∈ f, is_open (j u),
+   from λ _ _, is_open_compl_iff.mpr is_closed_closure,
+  have hj₂ : vᶜ ⊆ ⋃ u ∈ f, j u,
+   by { rw [compl_subset_comm], simp_rw [compl_Union, compl_compl'], assumption },
+  have hc : is_compact vᶜ, from (is_closed_compl_iff.mpr ho).compact,
+  rcases hc.elim_finite_subcover_image hj₁ hj₂ with ⟨g, hg₁, hg₂, hg₃⟩,
+  let w := ⋂₀ g,
+  have hw₁ : w ∈ f, from sInter_mem_sets_of_finite hg₂ hg₁,
+  have hw₂ : closure (image2 ϕ w s) ⊆ v, begin
+    rw compl_subset_comm at hg₃,
+    simp_rw [compl_Union, compl_compl'] at hg₃,
+    calc closure (image2 ϕ w s)
+        ⊆ ⋂ u ∈ g, closure (image2 ϕ u s) :
+          subset_Inter (λ _, subset_Inter (λ hu, closure_mono
+            (image2_subset (sInter_subset_of_mem hu) subset.rfl)))
+    ... ⊆ v : hg₃,
+  end,
+  exact ⟨_, hw₁, hw₂⟩,
+end
+
+/-- The ω-limit of a nonempty set w.r.t. a nontrivial filter is nonempty. -/
+lemma nonempty_omega_limit [compact_space β]
+  {s : set α} (hs : s.nonempty) (hf : ne_bot f) : (ω f ϕ s).nonempty :=
+begin
+  have : ω f ϕ s = ⋂ (u : ↥f.sets), closure (image2 ϕ u s), from bInter_eq_Inter _ _,
+  rw this,
+  apply is_compact.nonempty_Inter_of_directed_nonempty_compact_closed,
+  { rintro ⟨u₁, hu₁⟩ ⟨u₂, hu₂⟩,
+    use ⟨u₁ ∩ u₂, inter_mem_sets hu₁ hu₂⟩, split,
+    all_goals { exact closure_mono (image2_subset (by simp) subset.rfl) }},
+  { exact λ u, (nonempty.image2 (hf.nonempty_of_mem u.2) hs).mono subset_closure },
+  { exact λ _, is_closed_closure.compact },
+  { exact λ _, is_closed_closure },
+end
+
+end omega_limit
+
+/-!
+### Semiflows and flows
+
+A semiflow (resp. flow) on a topological space `α` by a topological
+semigroup (resp. monoid) `τ` is a continuous semigroup-act
+(resp. monoid-act) of `τ` on `α`.
+
+Anticipating the cases where `τ` is one of `ℕ`, `ℤ`, `ℝ⁺`, or `ℝ`, we
+use additive notation for the (semi)groups, though the definition does
+not require commutativity.
+-/
+
+/-- A semiflow on a topological space `α` by an a additive topological
+    semigroup `τ` is a continuous semigroup-act of `τ` on `α`.-/
+structure semiflow
+  (τ : Type*) [topological_space τ] [add_semigroup τ] [has_continuous_add τ]
+  (α : Type*) [topological_space α] :=
```

Comment:
Something that might become a small annoyance with possibly turning `flow` something that extends `has_vadd` is that the order of addition in `map_add` goes the other way around. This was a contrivance so that I could write `ϕ t` rather than `λ x, ϕ x t` in a lot of places. Either way, I think this shouldn't be hard to edit away in a prospective PR, so I'll leave it the way it is for now.

Also, I think you're right that working with `with_zero τ` is probably a better idea than maintaining a separate structure. will refactor.

Reply:
Extending `has_vadd` here requires a relatively large refactor of `mul_action` and `add_action`, so I don't think that you should do it before you'll become more familiar with internals of `mathlib`.